storage/src/test/java/org/apache/kafka/storage/internals/log/LogConcurrencyTest.java

```diff
@@ -0,0 +1,237 @@
+            final ConsumerTask consumer = new ConsumerTask(log, maxOffset);
+            final LogAppendTask appendTask = new LogAppendTask(log, maxOffset);
+
+            final Future<?> consumerFuture = executor.submit(consumer);
+            final Future<?> fetcherTaskFuture = executor.submit(appendTask);
+
+            fetcherTaskFuture.get();
+            consumerFuture.get();
+
+            validateConsumedData(log, consumer.getConsumedBatches());
+        } finally {
+            executor.shutdownNow();
+        }
+    }
+
+    private UnifiedLog createLog() throws IOException {
+        return createLog(new LogConfig(new Properties()));
+    }
+
+    private UnifiedLog createLog(LogConfig config) throws IOException {
+        log = UnifiedLog.create(
+            logDir,
+            config,
+            0L,
+            0L,
+            scheduler,
+            brokerTopicStats,
+            Time.SYSTEM,
+            5 * 60 * 1000,
+            new ProducerStateManagerConfig(TransactionLogConfig.PRODUCER_ID_EXPIRATION_MS_DEFAULT, false),
+            TransactionLogConfig.PRODUCER_ID_EXPIRATION_CHECK_INTERVAL_MS_DEFAULT,
+            new LogDirFailureChannel(10),
+            true,
+            Optional.empty()
+        );
+
+        return log;
+    }
+
+    private void validateConsumedData(UnifiedLog log, List<FetchedBatch> consumedBatches) {
+        final Iterator<FetchedBatch> iter = consumedBatches.iterator();
+        log.logSegments().forEach(segment ->
+            segment.log().batches().forEach(batch -> {
+                if (iter.hasNext()) {
+                    final FetchedBatch consumedBatch = iter.next();
+                    try {
+                        assertEquals(batch.partitionLeaderEpoch(),
+                            consumedBatch.epoch(), "Consumed batch with unexpected leader epoch");
+                        assertEquals(batch.baseOffset(),
+                            consumedBatch.baseOffset(), "Consumed batch with unexpected base offset");
+                    } catch (Throwable t) {
+                        fail("Consumed batch " + consumedBatch +
+                            " does not match next expected batch in log " + batch, t);
+                    }
+                }
+            })
+        );
+    }
+
+    /**
+     * Simple consumption task which reads the log in ascending order and collects
+     * consumed batches for validation
+     */
+    private static class ConsumerTask implements Callable<Void> {
+        private final UnifiedLog log;
+        private final int lastOffset;
+        private final List<FetchedBatch> consumedBatches = new ArrayList<>();
+
+        public ConsumerTask(UnifiedLog log, int lastOffset) {
+            this.log = log;
+            this.lastOffset = lastOffset;
+        }
+
+        public List<FetchedBatch> getConsumedBatches() {
+            return consumedBatches;
+        }
+
+        @Override
+        public Void call() throws Exception {
```

Comment:
the method `getConsumedBatches` could be removed if `call` returns `List<FetchedBatch>`